app/frontend/wallet/shelley-wallet.ts

```diff
@@ -420,12 +428,116 @@ const ShelleyWallet = ({
     return blockchainExplorer.getTxHistory([...base, ...legacy, account])
   }
 
+  async function getDelegationHistory(
+    shelleyAccountInfo,
+    validStakepools
+  ): Promise<DelegetionHistoryObject[]> {
+    const {delegations, rewards, withdrawals} = await blockchainExplorer.getDelegationHistory(
+      shelleyAccountInfo.accountPubkeyHex
+    )
+
+    const extractUrl = (poolHash) => {
+      const stakePool = validStakepools[poolHash]
+      if (stakePool) {
+        return stakePool.url
+      } else {
+        return null
+      }
+    }
+
+    const poolMetaUrls = distinct([
+      ...distinct(delegations.map((delegation) => extractUrl(delegation.poolHash))),
+      ...distinct(rewards.map((reward) => extractUrl(reward.poolHash))),
+    ]).filter((x) => x != null)
+
+    // Run requests for meta data in parallel
+    const metaDataPromises = poolMetaUrls.map((url: string) => ({
+      url,
+      metaDataPromise: getPoolInfo(url),
+    }))
+    const metaUrlToPoolNameMap = {}
+    for (const promise of metaDataPromises) {
+      const metaData = await promise.metaDataPromise
+      metaUrlToPoolNameMap[promise.url] = metaData.name
+    }
```

Comment:
That would make the function "poolHashToPoolName" async and if we would use it in deeper calls, it would eventually end up in map function, in which async function can't be called, it could be probably be done somehow, but it would complicate things more then simplify them.

Reply:
Resolved

Reply:
I changed it to more functional approach.